llama_index/tools/ondemand_loader_tool.py

```diff
@@ -0,0 +1,119 @@
+"""Ad-hoc data loader tool.
+
+Tool that wraps any data loader, and is able to load data on-demand.
+
+"""
+
+
+from llama_index.tools.types import BaseTool, ToolMetadata
+from llama_index.readers.base import BaseReader
+from typing import Any, List, Optional, Dict, Type, Tuple, Callable
+from llama_index.indices.base import BaseGPTIndex
+from llama_index.indices.vector_store import GPTVectorStoreIndex
+from pydantic import BaseModel, create_model
+from inspect import signature
+
+
+def create_schema_from_function(
+    name: str,
+    func: Callable[..., Any],
+    additional_fields: Optional[List[Tuple[str, Type, Any]]] = None,
+) -> Type[BaseModel]:
+    """Create schema from function."""
+    # NOTE: adapted from langchain.tools.base
+    fields = {}
+    params = signature(func).parameters
+    for param_name in params.keys():
+        param_type = params[param_name].annotation
+        param_default = params[param_name].default
+        if param_default is params[param_name].empty:
+            param_default = None
+        fields[param_name] = (param_type, param_default)
+
+    additional_fields = additional_fields or []
+    for field_name, field_type, field_default in additional_fields:
+        fields[field_name] = (field_type, field_default)
+
+    return create_model(name, **fields)  # type: ignore
+
+
+class OnDemandLoaderTool(BaseTool):
+    """On-demand data loader tool.
+
+    Loads data with query interface, stores in index, and queries
+    for relevant data with a natural language query string.
+
+    """
+
+    def __init__(
+        self,
+        reader: BaseReader,
+        index_cls: Type[BaseGPTIndex],
+        index_kwargs: Dict,
+        metadata: ToolMetadata,
+        use_query_str_in_loader: bool = False,
+        query_str_kwargs_key: str = "query_str",
```

Comment:
these param names not the most intuitive.

Reply:
yeah...will iterate on this probably